kotlinx-coroutines-core/common/src/sync/Mutex.kt

```diff
@@ -117,307 +131,138 @@ public suspend inline fun <T> Mutex.withLock(owner: Any? = null, action: () -> T
-                val update: Any = if (atomicOp.isDecided) EMPTY_UNLOCKED else atomicOp // restore if was already decided
-                (affected as MutexImpl)._state.compareAndSet(this, update)
-                return null // ok
-            }
-        }
-
-        override fun prepare(op: AtomicOp<*>): Any? {
-            val prepare = PrepareOp(op)
-            if (!mutex._state.compareAndSet(EMPTY_UNLOCKED, prepare)) return LOCK_FAIL
-            return prepare.perform(mutex)
-        }
-
-        override fun complete(op: AtomicOp<*>, failure: Any?) {
-            val update = if (failure != null) EMPTY_UNLOCKED else {
-                if (owner == null) EMPTY_LOCKED else Empty(owner)
-            }
-            mutex._state.compareAndSet(op, update)
-        }
-    }
-
-    public override fun holdsLock(owner: Any) =
-            _state.value.let { state ->
-                when (state) {
-                    is Empty -> state.locked === owner
-                    is LockedQueue -> state.owner === owner
-                    else -> false
-                }
-            }
 
     override fun unlock(owner: Any?) {
-        _state.loop { state ->
-            when (state) {
-                is Empty -> {
-                    if (owner == null)
-                        check(state.locked !== UNLOCKED) { "Mutex is not locked" }
-                    else
-                        check(state.locked === owner) { "Mutex is locked by ${state.locked} but expected $owner" }
-                    if (_state.compareAndSet(state, EMPTY_UNLOCKED)) return
-                }
-                is OpDescriptor -> state.perform(this)
-                is LockedQueue -> {
-                    if (owner != null)
-                        check(state.owner === owner) { "Mutex is locked by ${state.owner} but expected $owner" }
-                    val waiter = state.removeFirstOrNull()
-                    if (waiter == null) {
-                        val op = UnlockOp(state)
-                        if (_state.compareAndSet(state, op) && op.perform(this) == null) return
-                    } else {
-                        if ((waiter as LockWaiter).tryResumeLockWaiter()) {
-                            state.owner = waiter.owner ?: LOCKED
-                            waiter.completeResumeLockWaiter()
-                            return
-                        }
-                    }
-                }
-                else -> error("Illegal state $state")
-            }
+        while (true) {
+            // Is this mutex locked?
+            check(isLocked) { "This mutex is not locked" }
+            // Read the owner, waiting until it is set in a spin-loop if required.
+            val curOwner = this.owner.value
+            if (curOwner === NO_OWNER) continue // <-- ATTENTION, BLOCKING PART HERE
+            // Check the owner.
+            check(curOwner === owner) { "This mutex is locked by $curOwner, but $owner is expected" }
```

Comment:
This breaks some existing and documented behavior: see #2796. Please also add a test for this.

Reply:
Fixed